### PR TITLE
feat(host-preflights): add cidr preflight checks

### PR DESCRIPTION
**Pull request by JGAntunes**
#### What this PR does / why we need it:

Adds host preflight checks for the pod-cidr, service-cidr and cidr flags. If the `--pod-cidr`or `--service-cidr` checks are provided then those are the preflights that will be put in place, else we're gonna setup a preflight for wtv is provided through the `--cidr` flag.

This also fixes the current behaviour for CIDR flags where both service and a pod CIDR must be provided in order for us to use that VS the default CIDR flag.

#### Which issue(s) this PR fixes:

https://app.shortcut.com/replicated/story/113111/new-preflight-check-to-detect-whether-cluster-network-might-conflict-with-enterprise-s-internal-network

#### Does this PR require a test?

Added a series of tests.

Plus some manual executions:
```
root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --license local-dev/Joao.yaml --cidr 172.17.0.0/16
✔  Host files materialized!
✗  2 host preflights failed

 •  NTP is enabled but the system clock is not synchronized. Synchronize the system clock to continue.
 •  172.17.0.0/16 is not available. Use --cidr to specify a CIDR block of available private IP addresses (/16 or larger).

Please address these issues and try again.

root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --license local-dev/Joao.yaml --cidr 172.17.0.0/16
✔  Host files materialized!
✗  1 host preflight failed

 •  172.17.0.0/16 is not available. Use --cidr to specify a CIDR block of available private IP addresses (/16 or larger).

Please address this issue and try again.

root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --license local-dev/Joao.yaml --pod-cidr 172.17.0.0/16
✔  Host files materialized!
✗  1 host preflight failed

 •  172.17.0.0/16 is not available. Use --pod-cidr to specify an available CIDR block.

Please address this issue and try again.

root@node0:/replicatedhq/embedded-cluster# output/bin/embedded-cluster install run-preflights --license local-dev/Joao.yaml --service-cidr 172.17.0.0/16
✔  Host files materialized!
✗  2 host preflights failed

 •  NTP is enabled but the system clock is not synchronized. Synchronize the system clock to continue.
 •  172.17.0.0/16 is not available. Use --service-cidr to specify an available CIDR block.

Please address these issues and try again.

```

#### Does this PR require a release note?

```release-note
Adds preflight checks for the required subnets.
```

#### Does this PR require documentation?

NONE

### CODE REVIEW

File: pkg/preflights/host-preflight.yaml [hunk 1]
```diff
@@ -780,3 +795,36 @@ spec:
...
+        checkName: Subnet Availability
+        collectorName: Subnet
+        exclude: '{{ eq .GlobalCIDR.CIDR "" }}'
+        outcomes:
+          - fail:
+            when: "no-subnet-available"
+            message: specified subnet is not available
+          - pass:
+            when: "a-subnet-is-available"
+            message: specified subnet is available
```
Comment (JGAntunes) [inline pkg/preflights/host-preflight.yaml:830]: @ajp-io could I get some 👀 on the text here.

File: pkg/preflights/host-preflight.yaml [hunk 2]
```diff
@@ -127,6 +127,21 @@ spec:
...
+    - subnetAvailable:
+        collectorName: Services Subnet
+        exclude: '{{ or (ne .GlobalCIDR.CIDR "")  (eq .ServiceCIDR.CIDR "") }}'
+        CIDRRangeAlloc: '{{ .ServiceCIDR.CIDR }}'
+        desiredCIDR: {{.ServiceCIDR.Size}}
+    - subnetAvailable:
+        collectorName: Subnet
+        exclude: '{{ eq .GlobalCIDR.CIDR "" }}'
+        CIDRRangeAlloc: '{{ .GlobalCIDR.CIDR }}'
+        desiredCIDR: {{.GlobalCIDR.Size}}
```
Comment (ajp-io) [inline pkg/preflights/host-preflight.yaml:144]: Looks like you're offline, but I have a couple questions about these. Maybe we can link up tomorrow. I'll make some edits to the wording below for now.

File: pkg/preflights/host-preflight.yaml [hunk 3]
```diff
@@ -127,6 +127,21 @@ spec:
         collectorName: 'wildcard-check'
         hostnames:
           - '*'
+    - subnetAvailable:
+        collectorName: Pods Subnet
+        exclude: '{{ or (ne .GlobalCIDR.CIDR "")  (eq .PodCIDR.CIDR "") }}'
```
Comment (sgalsaleh) [inline pkg/preflights/host-preflight.yaml:132]: `(ne .GlobalCIDR.CIDR "")` is unnecessary

File: pkg/preflights/host-preflight.yaml [hunk 4]
```diff
@@ -127,6 +127,21 @@ spec:
...
+        CIDRRangeAlloc: '{{ .PodCIDR.CIDR }}'
+        desiredCIDR: {{.PodCIDR.Size}}
+    - subnetAvailable:
+        collectorName: Services Subnet
+        exclude: '{{ or (ne .GlobalCIDR.CIDR "")  (eq .ServiceCIDR.CIDR "") }}'
```
Comment (sgalsaleh) [inline pkg/preflights/host-preflight.yaml:137]: `(ne .GlobalCIDR.CIDR "")` is unnecessary

File: cmd/embedded-cluster/network.go [hunk 5]
```diff
@@ -50,3 +50,11 @@ func DeterminePodAndServiceCIDRs(c *cli.Context) (string, string, error) {
 	}
 	return netutils.SplitNetworkCIDR(c.String("cidr"))
 }
+
+// getCIDRs returns the CIDRs in use based on the provided cli flags.
+func getCIDRs(c *cli.Context) (string, string, string) {
+	if c.IsSet("pod-cidr") && c.IsSet("service-cidr") {
```
Comment (sgalsaleh) [inline cmd/embedded-cluster/network.go:56]:
```suggestion
	if c.IsSet("pod-cidr") || c.IsSet("service-cidr") {
```

Comment (sgalsaleh) [reply cmd/embedded-cluster/network.go:56]: this needs to be an `||` as well: https://github.com/replicatedhq/embedded-cluster/blob/main/cmd/embedded-cluster/network.go#L48

File: pkg/preflights/host-preflight.yaml (hunk 2 above)
Comment (ajp-io) [reply pkg/preflights/host-preflight.yaml:144]: Talked with Salah and his comments cover my questions. I'm updating wording now.

File: cmd/embedded-cluster/network.go (hunk 5 above)
Comment (JGAntunes) [reply cmd/embedded-cluster/network.go]: Done via - https://github.com/replicatedhq/embedded-cluster/pull/1355/commits/106fee65e04a20a6c8fbd256ae9e8a1bd29a402b

File: pkg/preflights/host-preflight.yaml (hunk 4 above)
Comment (JGAntunes) [reply pkg/preflights/host-preflight.yaml]: Done via - https://github.com/replicatedhq/embedded-cluster/pull/1355/commits/51a58ef43c7df77b23c5c10c820cf92fa0672dba

File: pkg/preflights/host-preflight.yaml (hunk 3 above)
Comment (JGAntunes) [reply pkg/preflights/host-preflight.yaml]: Done via - https://github.com/replicatedhq/embedded-cluster/pull/1355/commits/51a58ef43c7df77b23c5c10c820cf92fa0672dba

File: cmd/embedded-cluster/network_test.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import (
+	"flag"
+	"testing"
+
+	"github.com/k0sproject/k0s/pkg/apis/k0s/v1beta1"
+	"github.com/stretchr/testify/require"
+	"github.com/urfave/cli/v2"
+)
+
+func Test_getCIDRs(t *testing.T) {
```
Comment (sgalsaleh) [inline cmd/embedded-cluster/network_test.go:12]: don't see a use case for when only `cidr` flag is used, but won't block pr on it

**Review by sgalsaleh — DISMISSED** [review]

File: cmd/embedded-cluster/network_test.go (hunk 6 above)
Comment (JGAntunes) [reply cmd/embedded-cluster/network_test.go:12]: Done via -https://github.com/replicatedhq/embedded-cluster/pull/1355/commits/6febab9af8c35aa78351c089f465efec0ab434d4

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**